lib/p2psocket.js

```diff
@@ -77,93 +101,200 @@ P2PSocket.prototype.parseInv = function(payload) {
+   */
+  if ( content.version > 1 ) {
+    var coinbase = content.tx[0];
+    content.height = 0;
+
+    var script = new Buffer(coinbase.vin[0].coinbase,'hex');
+    var s = script.readUInt8(0);
+    for (i=0; i<s && i<3; i++) {
+      content.height += script.readUInt8(i+1) * ( i > 0 ? i*256 : 1);
+    }
+  }
+
+
+  return content;
+}
+
 P2PSocket.prototype.parseTx = function(payload) {
+
   var content = {
     version: payload.readUInt32LE(0),
+    hex:  null,
+    txid: null,
+    lock_time: null,
     vin:  [],
     vout: [],
   };
-  
+
   var i = 4;
-  
+
   var cin = payload.readVarInt(i);
   i += cin.size;
-  
+
   var n;
   for ( n=0; n < cin.value; n++ ) {
-  
+
     var tx = {
-      previous_output: {
-        txid: payload.slice(i, i+32),
-        vout: payload.readUInt32LE(i+32),
-      },
+      txid: inPlaceReverse(payload.slice(i, i+32),true).toString('hex'),
+      vout: payload.readUInt32LE(i+32)
     };
-    
+
     i += 36;
-    
+
     var sl = payload.readVarInt(i);
     i += sl.size;
-    
-    tx.scriptSig = payload.slice(i, i+sl.value);
+
+    var scriptSig = payload.slice(i, i+sl.value);
+    tx.scriptSig = {
+      hex: scriptSig.toString('hex'),
+      asm: scriptToAsm(scriptSig),
+    }
+
     i += sl.value;
-    
+
     tx.sequence = payload.readUInt32LE(i);
     i += 4;
-    
-    content.vin.push(tx);      
+
+    if ( tx.vout === tx.sequence && tx.txid.match(/^0+$/) ) {
+      tx = {
+        coinbase: tx.scriptSig.hex,
+        sequence: tx.sequence,
+      }
+    }
+
+    content.vin.push(tx);
   }
 
   var cout = payload.readVarInt(i);
   i += cout.size;
-  
+
   for ( n=0; n < cout.value; n++ ) {
-  
+
     var tx = {
-      value: payload.readInt64(i),
+      value: payload.readInt64(i) / this._divisor,
     };
-    
+
     i += 8;
-    
+
     var pkl = payload.readVarInt(i);
     i += pkl.size;
-    
-    tx.scriptPubKey = payload.slice(i, i+pkl.value);
+
+    var scriptPubKey = payload.slice(i, i+pkl.value);
+    tx.scriptPubKey = {
+      hex: scriptPubKey.toString('hex'),
+      asm: scriptToAsm(scriptPubKey),
+    }
+
+    if ( scriptPubKey.length === 25 && scriptPubKey[0] === opcodes.OP_DUP &&
+         scriptPubKey[1] === opcodes.OP_HASH160 && scriptPubKey[2] === 20 &&
+         scriptPubKey[23] === opcodes.OP_EQUALVERIFY &&
+         scriptPubKey[24] === opcodes.OP_CHECKSIG
+     ) {
+      tx.scriptPubKey.reqSigs = 1;
+      tx.scriptPubKey.type = 'pubkeyhash';
+
+      var hash = scriptPubKey.slice( 3, 23 );
+
+      var addrBuf = new Buffer( hash.length + 1 );
+      addrBuf[0] = this._p2pkh_byte;
+      hash.copy( addrBuf, 1, 0 );
+
+      var sha256_pass1 = crypto.createHash('sha256').update(addrBuf).digest();
+      var sha256 = crypto.createHash('sha256').update(sha256_pass1).digest();
+
+      tx.scriptPubKey.addresses = [
+        bs58.encode( Buffer.concat( [addrBuf, sha256.slice(0,4)], addrBuf.length + 4 ) )
+      ];
+    }
```

Comment:
This PR does not have P2SH decoding in it at the moment. That will be a separate PR.